### PR TITLE
Update of TTD code with various bug fixes and improvements for Node/VSCode debugging.

### PR DESCRIPTION
**Pull request by mrkmarron**
Clean up of inflate code.
Add suppression of getter execution when in debug mode.
Update TTD to specialized allocator.
Add support for Node/VSCode TTD.
Added partial JustMyCode mode.
Fixed various bugs.

### CODE REVIEW

File: bin/ch/ch.cpp [hunk 1]
```diff
@@ -240,16 +241,33 @@ HRESULT RunScript(const char* fileName, LPCWSTR fileContents, BYTE *bcBuffer, ch
...
-                IfJsErrorFailLog(ChakraRTInterface::JsTTDPrepContextsForTopLevelEventMove(chRuntime, nextEventTime, &snapEventTime));
+                bool needFreshCtxs = false;
+                JsErrorCode error = ChakraRTInterface::JsTTDGetSnapTimeTopLevelEventMove(chRuntime, moveMode, &nextEventTime, &needFreshCtxs, &snapEventTime, nullptr);
 
-                ChakraRTInterface::JsTTDMoveToTopLevelEvent(snapEventTime, nextEventTime);
+                if(error != JsNoError)
+                {
+                    if(error == JsErrorCategoryUsage)
+                    {
+                        wprintf(_u("Start time not in log range."));
```
Comment (obastemur) [inline bin/ch/ch.cpp]: both this and the one below misses `\n`. Was that intended?

File: bin/ch/ch.cpp [hunk 2]
```diff
@@ -404,7 +422,8 @@ HRESULT ExecuteTest(const char* fileName)
         ChakraRTInterface::JsTTDSetIOCallbacks(runtime, &Helpers::GetTTDDirectory, &Helpers::TTInitializeForWriteLogStreamCallback, &Helpers::TTGetLogStreamCallback, &Helpers::TTGetSnapshotStreamCallback, &Helpers::TTGetSrcCodeStreamCallback, &Helpers::TTReadBytesFromStreamCallback, &Helpers::TTWriteBytesToStreamCallback, &Helpers::TTFlushAndCloseStreamCallback);
 
         JsContextRef context = JS_INVALID_REFERENCE;
-        IfJsErrorFailLog(ChakraRTInterface::JsTTDCreateContext(runtime, &context));
+        IfJsErrorFailLog(ChakraRTInterface::JsCreateContext(runtime, &context));
+        //IfJsErrorFailLog(ChakraRTInterface::JsTTDCreateContext(runtime, &context));
```
Comment (obastemur) [inline bin/ch/ch.cpp]: nit: remove?

File: lib/Jsrt/ChakraDebug.h [hunk 3]
```diff
@@ -893,36 +925,120 @@
...
+            _In_ JsValueRef buffer,
+            _In_ UINT32 index,
+            _In_ UINT32 count);
+
+    /// <summary>
+    ///     TTD API -- may change in future versions:
+    ///     Get info for notifying the TTD system that a raw buffer it shares with the host has been modified.
+    /// </summary>
+    /// <param name="instance">The array buffer we want to monitor for contents modification.</param>
+    /// <param name="initialPos">The first position in the buffer that may be modified.</param>
```
Comment (agarwal-sandeep) [inline lib/Jsrt/ChakraDebug.h]: initialModPos

File: lib/Jsrt/ChakraDebug.h [hunk 4]
```diff
@@ -893,36 +925,120 @@
...
+        _Out_ INT64* targetStartSnapTime,
+        _Out_opt_ INT64* targetEndSnapTime);
+
+    /// <summary>
+    ///     TTD API -- may change in future versions:
+    ///     Before calling JsTTDMoveToTopLevelEvent (which inflates a snapshot and replays) check to see if we want to reset the script context.
+    ///     We reset the script context if the move will require inflating from a different snapshot that the last one.
+    /// </summary>
+    /// <param name="runtimeHandle">The runtime handle that the script is executing in.</param>
+    /// <param name="createFreshContexts">Indicates if new script contexts need to be created for this move.</param>
```
Comment (agarwal-sandeep) [inline lib/Jsrt/ChakraDebug.h]: createFreshCtxs

File: lib/Jsrt/ChakraDebug.h [hunk 5]
```diff
@@ -893,36 +925,120 @@
...
-            _Out_ INT64* targetStartSnapTime);
+            _In_ bool createFreshCtxs);
+
+    /// <summary>
+    ///     TTD API -- may change in future versions:
+    ///     During debug operations some additional information is populated during replay. This runs the code between the given 
+    ///     snapshots to poulate this information which may be needed by the debugger to determine time-travel jump targets.
+    /// </summary>
+    ///<param name = "startSnapTime">The snapshot time that we will start executing from.< / param>
+    ///<param name = "startEndTime">The snapshot time that we will stop at (or -1 if we want to run to the end).< / param>
```
Comment (agarwal-sandeep) [inline lib/Jsrt/ChakraDebug.h]: endSnapTime, please other parameters documentation as well

File: lib/Jsrt/Jsrt.cpp [hunk 6]
```diff
@@ -3515,47 +3518,189 @@ CHAKRA_API JsTTDNotifyYield()
...
+
+        return JsNoError;
+    });
+
+    //We need to root add ref so we can find this during replay!!!
+    if(addRefObj != nullptr)
     {
-        scriptContext->GetThreadContext()->TTDLog->RecordJsRTEventLoopYieldPoint(scriptContext);
+        JsAddRef(addRefObj, nullptr);
     }
```
Comment (agarwal-sandeep) [inline lib/Jsrt/Jsrt.cpp]: Move this inside if check

File: lib/Jsrt/ChakraDebug.h [hunk 7]
```diff
@@ -92,7 +92,11 @@
         /// <summary>
         ///     Perform a single step over after a debug break if the next statement is a function call, else behaves as a stepin.
         /// </summary>
-        JsDiagStepTypeStepOver = 2
+        JsDiagStepTypeStepOver = 2,
+        /// <summary>  
+        ///     Perform a single step back to the previous statement.  
+        /// </summary>  
+        JsDiagStepTypeStepBack = 3
```
Comment (agarwal-sandeep) [inline lib/Jsrt/ChakraDebug.h]: Should say that this is only applicable for TTD

File: lib/Jsrt/JsrtDiag.cpp [hunk 8]
```diff
@@ -427,6 +427,18 @@ CHAKRA_API JsDiagSetStepType(
...
+#if ENABLE_TTD
+            ThreadContext* threadContext = runtime->GetThreadContext();
+
+            TTD::TTDebuggerSourceLocation bpLocation;
+            threadContext->TTDLog->GetPreviousTimeAndPositionForDebugger(bpLocation);
+            threadContext->TTDLog->SetPendingTTDBPInfo(bpLocation);
+
+            jsrtDebugManager->SetResumeType(BREAKRESUMEACTION_CONTINUE);
+#endif
+        }
```
Comment (agarwal-sandeep) [inline lib/Jsrt/JsrtDiag.cpp]: Should return error if TTD is not enabled

File: lib/Runtime/Debug/DebugDocument.cpp [hunk 9]
```diff
@@ -337,4 +337,25 @@ namespace Js
...
+        if(url == nullptr || wcslen(url) < 2)
+        {
+            return false;
+        }
+
+#ifdef WIN32
+        return ((url[1] == _u(':')) && (wcsstr(url, _u("\\node_modules\\")) == nullptr));
+#else
+        return ((url[0] == _u('/')) && (wcsstr(url, _u("/node_modules/")) == nullptr));
+#endif
```
Comment (agarwal-sandeep) [inline lib/Runtime/Debug/DebugDocument.cpp]: Even for debugging/test I don't prefer these kind of hard coded hacks, make code look bad. Please redesign

File: lib/Runtime/Library/SimdFloat32x4Lib.cpp [hunk 10]
```diff
@@ -942,12 +943,13 @@ namespace Js
                 JavascriptError::ThrowTypeError(scriptContext, JSERR_NeedNumber, _u("Lane index"));
             }
 
-            Var lane0 = args[3];
-            Var lane1 = args[4];
-            Var lane2 = args[5];
-            Var lane3 = args[6];
+            Var lanes[4];
+            for (uint i = 0; i < 4; ++i)
```
Comment (agarwal-sandeep) [inline lib/Runtime/Library/SimdFloat32x4Lib.cpp]: Are Simd*.h/cpp changes intentional?

**Comment by mrkmarron** [comment]
Superseded by PR #1338.

**Pull request closed without merging**